// @polkadot-api/sdk-governance/dist/esm/referenda/chainConfig.mjs
const SpenderOrigin = {
  Treasurer: "Treasurer",
  SmallTipper: "SmallTipper",
  BigTipper: "BigTipper",
  SmallSpender: "SmallSpender",
  MediumSpender: "MediumSpender",
  BigSpender: "BigSpender"
};
const originToTrack = {
  Treasurer: "treasurer",
  SmallTipper: "small_tipper",
  BigTipper: "big_tipper",
  SmallSpender: "small_spender",
  MediumSpender: "medium_spender",
  BigSpender: "big_spender"
};
const DOT_UNIT = 10000000000n;
const polkadotSpenderOrigin = (value) => {
  if (value <= 250n * DOT_UNIT) return SpenderOrigin.SmallTipper;
  if (value <= 1000n * DOT_UNIT) return SpenderOrigin.BigTipper;
  if (value <= 10000n * DOT_UNIT) return SpenderOrigin.SmallSpender;
  if (value <= 100000n * DOT_UNIT) return SpenderOrigin.MediumSpender;
  if (value <= 1000000n * DOT_UNIT) return SpenderOrigin.BigSpender;
  if (value <= 10000000n * DOT_UNIT) return SpenderOrigin.Treasurer;
  return null;
};
const KSM_UNIT = 1000000000000n;
const kusamaSpenderOrigin = (value) => {
  if (value <= 1n * KSM_UNIT) return SpenderOrigin.SmallTipper;
  if (value <= 5n * KSM_UNIT) return SpenderOrigin.BigTipper;
  if (value <= 333n * KSM_UNIT) return SpenderOrigin.SmallSpender;
  if (value <= 3333n * KSM_UNIT) return SpenderOrigin.MediumSpender;
  if (value <= 33333n * KSM_UNIT) return SpenderOrigin.BigSpender;
  return SpenderOrigin.Treasurer;
};

export { kusamaSpenderOrigin, originToTrack, polkadotSpenderOrigin };
